packages/anchor-listener/src/listener.ts

```diff
@@ -0,0 +1,131 @@
+import type { SupportedNetwork } from '@ceramicnetwork/anchor-utils'
+import type { Block, Provider } from '@ethersproject/providers'
+import {
+  type Observable,
+  type OperatorFunction,
+  type RetryConfig,
+  concatMap,
+  filter,
+  fromEvent,
+  map,
+  pipe,
+  scan,
+} from 'rxjs'
+
+import { type BlockWithAnchorProofs, loadAnchorProofs, mapLoadBlocks } from './loader.js'
+
+export type BlockEvent = BlockWithAnchorProofs & {
+  reorganized: false
+}
+
+export type ReorganizedBlockEvent = BlockWithAnchorProofs & {
+  reorganized: true
+  expectedParentHash: string
+}
+
+export type BlockListenerEvent = BlockEvent | ReorganizedBlockEvent
+
+export type ListenerParams = {
+  chainId: SupportedNetwork
+  confirmations: number
+  expectedParentHash?: string
+  provider: Provider
+  retryConfig?: RetryConfig
+}
+
+type OrderingState = { continuous: Array<number>; latestContinuous?: number; maxFuture?: number }
+
+export function createContinuousBlocksListener(
+  provider: Provider,
+  confirmations = 0
+): Observable<Array<number>> {
+  return fromEvent(provider, 'block').pipe(
+    // Map to latest block number minus wanted confirmations
+    map((blockNumber) => (blockNumber as number) - confirmations),
+    // Order continuous block numbers, keeping track of highest block received
+    scan<number, OrderingState, OrderingState>(
+      (state, blockNumber) => {
+        if (state.latestContinuous != null) {
+          if (blockNumber <= state.latestContinuous) {
+            // Already processed block number
+            return { ...state, continuous: [] }
+          }
+          if (blockNumber > state.latestContinuous + 1) {
+            // Future block number, keep track
+            return {
+              ...state,
+              continuous: [],
+              maxFuture: Math.max(state.maxFuture ?? 0, blockNumber),
+            }
+          }
+        }
+
+        return state.maxFuture == null
+          ? // Load single block
+            { continuous: [blockNumber], latestContinuous: blockNumber }
+          : // Load range from new block number to max tracked future block number
+            {
+              continuous: new Array(state.maxFuture - blockNumber + 1).fill(0).map((_, i) => {
+                return blockNumber + i
+              }),
+              latestContinuous: state.maxFuture,
+            }
+      },
+      { continuous: [] }
+    ),
+    // Extract continous blocks from ordering state
+    map((state) => state.continuous),
+    // Only push non-empty lists of blocks
+    filter((blocks) => blocks.length !== 0)
+  )
+}
+
+type ProcessingState = { status: 'process' | 'reorganized'; previousHash: string; block: Block }
+type ProcessingSeed = { status: 'initial' }
+
+export function mapProcessBlock(
```

Comment:
do all these functions in this file and in loader.ts all need to be exported?  Can we either wrap them in a class with a small public API vs private helper methods, or do something else to make it clearer which functions are really intended to be consumed by users of this package vs which are mostly internal?

Reply:
All exported functions should be safe to use by consumers as needed, there's no need to make anything "private".
We don't have to export them all but I guess I don't see the downside, in case they can be useful for the consumer logic?

Reply:
I've always believed it is good engineering best practice to leave APIs as small and narrow as possible to achieve the needed goals.  Putting functionality into the public API "just in case someone needs it" increases the surface area that needs to be maintained and makes it harder to refactor in the future.  I'd imagined one single very simple function call at the top level that you point at a contract address (possibly with some optional configuration like starting block number) and that then spits out merkle root CIDs.  Then everything about how it actually gets that information is an implementation detail.

We can always expose more functionality later if someone needs it, it's a lot harder to take functionality away once people start relying on it.

Reply:
I agree for "high-level" APIs I prefer a smaller surface, but these packages are mainly low-level utilities that can be useful in different cases. If people use them, to me it's more a good sign that the library is useful to the community rather than a downside, no?
Anyway, I reduced the exports to the 3 highest level functions, as you wrote we can always export more if needed.